cpp/src/arrow/json/chunked-builder.h

```diff
@@ -0,0 +1,76 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+namespace internal {
+class TaskGroup;
+}  // namespace internal
+
+class Array;
+class MemoryPool;
+class DataType;
+class Field;
+class ChunkedArray;
+
+namespace json {
+
+class PromotionGraph;
+
+class ARROW_EXPORT ChunkedArrayBuilder {
+ public:
+  virtual ~ChunkedArrayBuilder() = default;
+
+  /// Spawn a task that will try to convert and insert the given JSON block
+  virtual void Insert(int64_t block_index,
+                      const std::shared_ptr<Field>& unconverted_field,
+                      const std::shared_ptr<Array>& unconverted) = 0;
+
+  /// Return the final chunked array.
+  /// chunk_lengths may be empty if all chunks were inserted, otherwise if the
+  /// nth chunk is missing then it is assumed to contain chunk_lengths[n] nulls
```

Comment:
Right, for example it would be possible for the fourth chunk of a JSON file to contain a new field. In that case the first three chunks would never be inserted for that field. `chunk_lengths` is only allowed empty to allow for the root builder (which *has* had all chunks inserted) and so its `chunked_lengths_` member is complete. I don't love this aspect of `ChunkedArrayBuilder` but I'm not sure how to resolve it; one option would be to insert absent chunks in `Finish()`.

Reply:
@pitrou I've removed `chunk_lengths`, how does it look?